core/base/composition.hpp

```diff
@@ -0,0 +1,159 @@
+                    public EnableCreateMethod<Composition<ValueType>> {
+    friend class EnablePolymorphicObject<Composition, LinOp>;
+    friend class EnableCreateMethod<Composition>;
+
+public:
+    using value_type = ValueType;
+
+protected:
+    /**
+     * Creates an empty operator composition (0x0 operator).
+     *
+     * @param exec  Executor associated to the composition
+     */
+    explicit Composition(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Composition>(exec)
+    {}
+
+    /**
+     * Creates a composition of operators using the operators in a range.
+     *
+     * @tparam Iterator  a class representing iterators over the
+     *                  perators of the linear combination
+     *
+     * @param begin  iterator pointing to the first operator
+     * @param end  iterator pointing behind the last operator
+     */
+    template <typename Iterator,
+              typename = xstd::void_t<
+                  typename std::iterator_traits<Iterator>::iterator_category>>
+    explicit Composition(Iterator begin, Iterator end)
+        : EnableLinOp<Composition>(first_or_throw(begin, end)->get_executor()),
+          operators_(begin, end)
+    {
+        this->set_size(gko::dim<2>{operators_.front()->get_size()[0],
+                                   operators_.back()->get_size()[1]});
+        for (size_type i = 1; i < operators_.size(); ++i) {
+            ASSERT_CONFORMANT(operators_[i - 1], operators_[i]);
+        }
+    }
+
+    /**
+     * Creates a composition of operators using the specified list of operators.
+     *
+     * @tparam Rest  types of trailing parameters
+     *
+     * @param oper  the first operator
+     * @param rest  remainging operators
+     */
+    template <typename... Rest>
+    explicit Composition(std::shared_ptr<const LinOp> oper, Rest &&... rest)
+        : Composition(std::forward<Rest>(rest)...)
+    {
+        ASSERT_CONFORMANT(oper, operators_[0]);
+        operators_.insert(begin(operators_), oper);
+        this->set_size(gko::dim<2>{operators_.front()->get_size()[0],
+                                   operators_.back()->get_size()[1]});
+    }
+
+    /**
+     * Creates a composition of operators using the specified list of operators.
+     *
+     * @param oper  the first operator
+     *
+     * @note this is the base case of the template constructor
+     *       Composition(std::shared_ptr<const LinOp>, Rest &&...)
+     */
+    explicit Composition(std::shared_ptr<const LinOp> oper)
+        : EnableLinOp<Composition>(oper->get_executor(), oper->get_size()),
+          operators_{oper}
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+private:
+    template <typename Iterator>
+    auto first_or_throw(Iterator begin, Iterator end) -> decltype(*begin)
```

Comment:
Since this is only a function due to restrictions of the language, I wouldn't do that.
What I really want to write is something like:

```
Constructor(begin, end) : 
  BaseConstructor(begin == end ? throw Exception() : something_with(*begin))
{}
```

which is of course not valid, since `throw` is a statement and not an expression.
Hence, I have to write the expression as an additional function.

Another reason why not share is that then the exception is thrown from some other file, while in fact it occurred in the constructor.

Reply:
Actually, maybe I could use a lambda to get around it...